### PR TITLE
Commission structure (WIP)

### PR DESCRIPTION
**Pull request by jamesmorgan**
* [x] Ability to define different structures per type (DIG/PYH)
* [x] Ability to update each type
* [x] Handle artwork of zero fee
* [x] Upgrade to open-zepllin 1.8.0
* [x] Lots of tests...does it work?

### CODE REVIEW

File: contracts/KnownOriginDigitalAsset.sol [hunk 1]
```diff
@@ -78,6 +83,11 @@ contract KnownOriginDigitalAsset is ERC721Token {
     curator = msg.sender;
     commissionAccount = _commissionAccount;
     contractDeveloper = _contractDeveloper;
+
+    // TODO work out how to do floating point division
```
Comment (jamesmorgan) [inline contracts/KnownOriginDigitalAsset.sol]: @andygray last few bits for tomo is to work out how to do floating point division

File: contracts/KnownOriginDigitalAsset.sol [hunk 2]
```diff
@@ -225,28 +211,38 @@ contract KnownOriginDigitalAsset is ERC721Token {
...
+  internal
+  {
+    bytes16 edition = tokenIdToEdition[_tokenId];
 
-      // send ether to owner instantly
-      curator.transfer(curatorTotal);
+    string memory typeCode = getTypeFromEdition(edition);
 
-      PurchasedWithEther(_tokenId, msg.sender);
+    // TODO handle not found?
```
Comment (jamesmorgan) [inline contracts/KnownOriginDigitalAsset.sol]: @andygray also make this diffensive, I am thinking a simply if statement with some hard defaults of 5% each.

File: test/truffle/KnownOriginDigitalAsset.test.js [hunk 3]
```diff
@@ -1245,11 +1245,146 @@ contract('KnownOriginDigitalAsset', function (accounts) {
...
-            from: curator
+            from: _curator
           }));
         });
       });
     });
 
   });
+
+  describe('commission structure', function () {
```
Comment (jamesmorgan) [inline test/truffle/KnownOriginDigitalAsset.test.js]: @andygray started a load of tests, covers most things but using simply non floating commissions

File: contracts/KnownOriginDigitalAsset.sol [hunk 4]
```diff
@@ -78,6 +83,11 @@ contract KnownOriginDigitalAsset is ERC721Token {
     curator = msg.sender;
     commissionAccount = _commissionAccount;
     contractDeveloper = _contractDeveloper;
+
+    // TODO work out how to do floating point division
+    // Setup default commission structures
+    tokenIdToCommission["DIG"] = CommissionStructure({curator : 12, developer : 12});
+    tokenIdToCommission["PHY"] = CommissionStructure({curator : 24, developer : 15});
```
Comment (jamesmorgan) [inline contracts/KnownOriginDigitalAsset.sol]: TODO - update to use correct floating point commissions

File: contracts/KnownOriginDigitalAsset.sol [hunk 5]
```diff
@@ -204,6 +166,30 @@ contract KnownOriginDigitalAsset is ERC721Token {
...
+    require(_curator > 0);
+    require(_developer > 0);
+    require((_curator + _developer) < 100);
+
+    tokenIdToCommission[_type] = CommissionStructure({curator: _curator, developer: _developer});
+    return true;
+  }
+
+  function getCommissionForType(string _type)
+  public
```
Comment (jamesmorgan) [inline contracts/KnownOriginDigitalAsset.sol]: @andygray  What you reckon about making this public or only viewable by management?

File: contracts/KnownOriginDigitalAsset.sol [hunk 6]
```diff
@@ -323,4 +319,61 @@ contract KnownOriginDigitalAsset is ERC721Token {
...
+
+  function getTypeFromEdition(bytes16 _bytes16) public pure returns (string){
+    bytes memory bytesArray = new bytes(3);
+    uint pos = 0;
+    for (uint256 i = 13; i < 16; i++) {
+      bytesArray[pos] = _bytes16[i];
+      pos++;
+    }
+    return string(bytesArray);
+  }
```
Comment (jamesmorgan) [inline contracts/KnownOriginDigitalAsset.sol]: @andygray I also reckon we add solidity docs to all methods where appropriate - good thing is that contract source on etherscan includes them as well - thoughts?

File: contracts/KnownOriginDigitalAsset.sol [hunk 7]
```diff
@@ -78,6 +85,10 @@ contract KnownOriginDigitalAsset is ERC721Token {
     curator = msg.sender;
     commissionAccount = _commissionAccount;
     contractDeveloper = _contractDeveloper;
```
Comment (andygray) [inline contracts/KnownOriginDigitalAsset.sol]: use `developer` like in commission structure

**Comment by andygray** [comment]
Lovely

**Pull request merged**